src/iface/fragmentation.rs

```diff
@@ -0,0 +1,555 @@
+#[derive(Debug)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub struct PacketAssemblerSet<'a, Key: Eq + Ord + Clone + Copy, Info: PacketAssemblerInfo> {
+    packet_buffer: ManagedSlice<'a, PacketAssembler<'a, Info>>,
+    index_buffer: ManagedMap<'a, Key, u8>,
+}
+
+impl<'a, K: Eq + Ord + Clone + Copy, Info: PacketAssemblerInfo> PacketAssemblerSet<'a, K, Info> {
+    /// Create a new set of packet assemblers.
+    ///
+    /// # Panics
+    ///
+    /// This will panic when:
+    ///   - The packet buffer and index buffer don't have the same size or are empty (when they are
+    ///   both borrowed).
+    ///   - The packet buffer is empty (when only the packet buffer is borrowed).
+    ///   - The index buffer is empty (when only the index buffer is borrowed).
+    pub fn new<FB, IB>(packet_buffer: FB, index_buffer: IB) -> Self
+    where
+        FB: Into<ManagedSlice<'a, PacketAssembler<'a, Info>>>,
+        IB: Into<ManagedMap<'a, K, u8>>,
+    {
+        let packet_buffer = packet_buffer.into();
+        let index_buffer = index_buffer.into();
+
+        match (&packet_buffer, &index_buffer) {
+            (ManagedSlice::Borrowed(f), ManagedMap::Borrowed(i)) => {
+                if f.len() != i.len() {
+                    panic!("The amount of places in the index buffer must be the same as the amount of possible fragments assemblers.");
+                }
+            }
+            #[cfg(any(feature = "std", feature = "alloc"))]
+            (ManagedSlice::Borrowed(f), ManagedMap::Owned(_)) => {
+                if f.is_empty() {
+                    panic!("The packet buffer cannot be empty.");
+                }
+            }
+            #[cfg(any(feature = "std", feature = "alloc"))]
+            (ManagedSlice::Owned(_), ManagedMap::Borrowed(i)) => {
+                if i.is_empty() {
+                    panic!("The index buffer cannot be empty.");
+                }
+            }
+            #[cfg(any(feature = "std", feature = "alloc"))]
+            (ManagedSlice::Owned(_), ManagedMap::Owned(_)) => (),
+        }
+
+        Self {
+            packet_buffer,
+            index_buffer,
+        }
+    }
+
+    /// Reserve a [`PacketAssembler`], which is linked to a specific key.
+    /// Returns the reserved fragments assembler.
+    ///
+    /// # Errors
+    ///
+    /// - Returns [`Error::PacketAssemblerSetFull`] when every [`PacketAssembler`] in the buffer is used (only
+    /// when the non allocating version of is used).
+    pub(crate) fn reserve_with_key(&mut self, key: &K) -> Result<&mut PacketAssembler<'a, Info>> {
+        if self.packet_buffer.len() == self.index_buffer.len() {
+            match &mut self.packet_buffer {
+                ManagedSlice::Borrowed(_) => return Err(Error::PacketAssemblerSetFull),
+                #[cfg(any(feature = "std", feature = "alloc"))]
+                ManagedSlice::Owned(b) => {
+                    b.resize_with(self.index_buffer.len() + 1, || {
```

Comment:
this could be a .push(), we're just pushing 1 elem.

Reply:
Changed it.